### PR TITLE
introduce new keyframe based animation to graphic component and custom series

### PR DESCRIPTION
**Pull request by pissang**
This pull request is in the type of:

- [ ] bug fixing
- [x] new feature
- [ ] others

## What does this PR do?

This PR includes two major features:

1. Introduces transition to graphic component. Which is similar to custom series.
2. New keyframe based animation in both graphic component and custom series.

## Transition in graphic component.

We introduced property transition in custom series in https://github.com/apache/echarts/pull/12775.

It allow us applying complex transform/shape/style transition to the elements when data is updated. Even more, we can use `enterFrom` and `leaveTo` to configure the `enter` and `leave` animation.

In this PR we also brings this flexible transition option to the `graphic` component. Here is an simple example:

```ts
// Display an orange circle at [50, 100]
chart.setOption({
  graphic: {
    elements: [{
      type: 'circle',
      left: 'left',
      y: 100,
      x: 50,
      // Enter from outside
      enterFrom: {
        x: -100
      },
      // Configure x and y property can be transition.
      transition: ['x', 'y'],
      shape: {
        cx: 0,
        cy: 0,
        r: 50
      },
      style: {
        fill: 'orange'
      }
    }]
  }
});
...
// Move to [200, 100] smoothly
chart.setOption({
  graphic: {
    elements: [{
      x: 200
    }]
  }
})
...
// Layout options like left/top/right/bottom is also supported.
// Move to [center, center] smoothly
chart.setOption({
  graphic: {
    elements: [{
      left: 'center',
      top: 'center'
    }]
  }
});
...
// Enlarge the circle
chart.setOption({
  graphic: {
    elements: [{
      transition: ['shape'],
      shape: {
        r: 100
      }
    }]
  }
});
// Transition the color to red.
chart.setOption({
  graphic: {
    elements: [{
      transition: ['style'],
      style: {
        color: 'red'
      }
    }]
  }
});
```

As we can see, the option keeps the same as the custom series. To make the transition more usable. We add enhancement:

1. Support `transition: 'all'` to configure all properties can have transition animation.

Enumerate the properties in the transition option can be tedious. We can now use the keyword `'all'` on the `transition` configuration if we don't wan't to care about which property can have the animation.

2. Can configure the nested properties include `style`, `shape`, `extra` in the top level `enterFrom` and `leaveTo`

For example:
```ts
enterFrom: {
  x: -100,
  style: {
    opacity: 0
  }
}
```

3. Animation configuration per element.

We provide `enterAnimation`, `updateAnimation`, `leaveAnimation` to configure the animation `duration`, `easing`, `delay` per element.

For example:
```ts
graphic: {
  elements: [{
    enterAnimation: {
      duration: 1000
    },
    updateAnimation: {
      duration: 500
    },
    leaveAnimation: {
      duration: 200
    }
  }]
}
```

All these three enhancements will be available in both graphic component and custom series.

## New keyframe animation API

Transition animation only applies when data is created, changed or removed. It only has start and end state and will only run once.

In order to providing more complex multiple keyframe looped animations. We introducing a brand new keyframe animation API to the custom series and graphic components in this PR.

We can use it to do things like:

1. Displaying an awesome loading animation.
2. Having background with some dynamic effect.
3. Creating your own animation effect in the chart. Not only the builtin ripple effect or trail effect.
4. Load a lottie animation file and added it in the echarts.
5. Or just have fun playing with the animation.

###  Interface

```ts
interface KeyframeAnimationKeyframe {
  // Easing of each keyframe
  easing?: AnimationEasing
  // Percent of this keyframe
  percent?: number

  ...// Element properties.
};

interface KeyframeAnimation {
  duration?: string
  easing?: AnimationEasing
  delay?: number
  loop?: boolean
  keyframes?: KeyframeAnimationKeyframe[]
}

interface GraphicElement {
  ... // Other properties
  keyframeAnimation?: KeyframeAnimation | KeyframeAnimation[]
}
```

#### Text animation

Code:
```ts
chart.setOption({
  graphic: {
    elements: [{
      type: 'text',
      x: 100,
      y: 100,
      style: {
        text: 'Apache ECharts',
        fontSize: 50,
        fontWeight: 'bold',
        lineDash: [0, 200],
        lineDashOffset: 0,
        fill: 'transparent',
        stroke: '#000',
        lineWidth: 1
      },
      keyframeAnimation: {
        duration: 3000,
        loop: true,
        keyframes: [{
          percent: 0.7,
          style: {
            fill: 'transparent',
            lineDashOffset: 200,
            lineDash: [200, 0]
          }
        }, {
          // Hold a bit.
          percent: 0.8,
          style: {
            fill: 'transparent'
          }
        }, {
          percent: 1,
          style: {
            fill: 'black'
          }
        }]
      },
    }]
  }
})
```

Result:

https://user-images.githubusercontent.com/841551/146493515-d8c24d24-23d1-4c6d-9ec4-f6dc715eabef.mov

#### Procedural wave animation

Code:

```ts
const elements = [];
for (let x = 20; x < width; x += 40) {
  for (let y = 20; y < height; y += 40) {
    elements.push({
      type: 'circle',
      x, y,
      style: {
        fill: '#800080'
      },
      shape: {
        r: 30
      },
      keyframeAnimation: {
        duration: 4000,
        loop: true,
        delay: (noise.perlin2(x / 500 + 0, y / 500 + 100) - 1) * 4000,
        keyframes: [{
          percent: 0.5,
          easing: 'sinusoidalInOut',
          style: {
              fill: '#ffa500'
          },
          scaleX: 0.1,
          scaleY: 0.1
        }, {
          percent: 1,
          easing: 'sinusoidalInOut',
          style: {
              fill: '#800080'
          },
          scaleX: 1,
          scaleY: 1
        }]
      }
    });
  }
}
chart.setOption({
  graphic: {
    elements
  }
})
```

Result:

https://user-images.githubusercontent.com/841551/146493565-0fb939d7-16f4-44bc-a461-ab748ac10296.mp4

#### Marker jump effect

Use keyframe based animation in the custom series to have more complex marker effect.

Code:
```ts
renderItem(params, api) {
  const coord = api.coord([api.value(0, params.dataIndex), api.value(1, params.dataIndex)]);

  const circles = [];
  for (let i = 0; i < 5; i++) {
    circles.push({
      type: 'circle',
      shape: {
        cx: 0,
        cy: 0,
        r: 30
      },
      style: {
        stroke: 'red',
        fill: 'none',
        lineWidth: 2
      },
      // Ripple animation
      keyframeAnimation: {
        duration: 4000,
        loop: true,
        delay: -i / 4 * 4000,
        keyframes: [{
          percent: 0,
          scaleX: 0,
          scaleY: 0,
          style: {
            opacity: 1
          }
        }, {
          percent: 1,
          scaleX: 1,
          scaleY: 0.4,
          style: {
            opacity: 0
          }
        }]
      }
    });
  }
  return {
    type: 'group',
    x: coord[0],
    y: coord[1],
    children: [
      ...circles,
      {
        type: 'path',
        shape: {
          d: 'M16 0c-5.523 0-10 4.477-10 10 0 10 10 22 10 22s10-12 10-22c0-5.523-4.477-10-10-10zM16 16c-3.314 0-6-2.686-6-6s2.686-6 6-6 6 2.686 6 6-2.686 6-6 6z',
          x: -10,
          y: -35,
          width: 20,
          height: 40
        },
        style: {
          fill: 'red'
        },
        // Jump animation.
        keyframeAnimation: {
          duration: 1000,
          loop: true,
          delay: Math.random() * 1000,
          keyframes: [{
            y: -10,
            percent: 0.5,
            easing: 'cubicOut'
          }, {
            y: 0,
            percent: 1,
            easing: 'bounceOut'
          }]
        }
      }
    ]
  }
```

Result:

https://user-images.githubusercontent.com/841551/146501306-644a2196-2750-4fa2-b394-763f0fd0c6ad.mov

#### Lottie animation

TBD

## Misc

- [x] The API has been changed (apache/echarts-doc#xxx).
- [x] This PR depends on ZRender changes (ecomfe/zrender#857).

### CODE REVIEW

File: src/animation/customGraphicKeyframeAnimation.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                return;
+            }
+
+            let propKeys = keys(kfValues);
+            if (!targetPropName) {
+                // PENDING performance?
+                propKeys = filter(
+                    propKeys, key => key !== 'percent' && key !== 'easing'
+                    && key !== 'shape' && key !== 'style' && key !== 'extra'
+                );
```
Comment (susiwen8) [inline src/animation/customGraphicKeyframeAnimation.ts:119]: It's better to use `include`?
```ts
!['percent', 'easing', 'shape', 'style', 'extra'].include(key)
```

File: src/animation/customGraphicKeyframeAnimation.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+const getStateToRestore = makeInner<StateToRestore, Element>();
+
+export interface ElementKeyframeAnimationOption<Props extends Record<string, any>> extends AnimationOption {
+    // Animation configuration for keyframe based animation.
+    loop?: boolean
+    keyframes?: AnimationKeyframe<Props>[]
+}
+
+/**
+ * Stopped previous keyframe animation and restore the attributes.
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/animation/customGraphicKeyframeAnimation.ts]: Fixed

File: src/animation/customGraphicKeyframeAnimation.ts (hunk 1 above)
Comment (pissang) [reply src/animation/customGraphicKeyframeAnimation.ts]: Thanks for the suggestion. Changed to using `indexOf` to check if key is included. `includes` is not supported in IE. Which we haven't dropped officialy yet.

**Comment by plainheart** [comment]
Seems the unit test didn't pass for some type errors.

**Comment by pissang** [comment]
UT is fixed

**Comment by plainheart** [comment]
I noticed the strange behavior when hovering on the animated marker. Is there anything wrong?

**Comment by pissang** [comment]
@plainheart Seems it's because keyframe animation is conflicted to the hover animation.

**Comment by pissang** [comment]
@plainheart Improved it a bit in the PR https://github.com/ecomfe/zrender/pull/859. Now the behavior is:

https://user-images.githubusercontent.com/841551/147042848-f2e7d60e-66a5-4cdf-a0a9-616bba0057c0.mov

**Review by plainheart — APPROVED** [review]
Looks good!

**Pull request merged**